### ISSUE TITLE
py.typed file is not applied recursively

### ISSUE THREAD

**Issue by kylegentle** [microsoft/pyright#9671]
**Describe the bug**
[PEP 561](https://peps.python.org/pep-0561/#packaging-type-information) specifies that py.typed applies recursively:

>Package maintainers who wish to support type checking of their code MUST add a marker file named py.typed to their package supporting typing. This marker applies recursively: if a top-level package includes it, all its sub-packages MUST support type checking as well.

It seems like pyright only respects py.typed on a per-module level.

**Code or Screenshots**
Repro layout:
```
.
├── example-py-pkg
│   ├── module
│   │   └── example.py
│   └── py.typed
├── foo.py
└── pyproject.toml
```

`pyproject.toml` contains:
```
[tool.pyright]
typeCheckingMode = "strict"
```

`module.example.py` defines a simple function `foo`:
```python
def func() -> int:
    return 2
```

`foo.py` contains:
```python
import module.example as example

example.func()
```

Run pyright:
```bash
$ PYTHONPATH=./example-py-pkg pyright foo.py
/path/to/foo.py
  /path/to/foo.py:1:8 - error: Stub file not found for "module.example" (reportMissingTypeStubs)
1 error, 0 warnings, 0 informations
```

If a `py.typed` file is added under `example-py-pkg/module/py.typed`, pyright passes:
```bash
$ touch example-py-pkg/module/py.typed
$ PYTHONPATH=./example-py-pkg pyright foo.py
0 errors, 0 warnings, 0 informations
```

I first ran into this problem when including a py.typed file at the subpackage level of a namespace package, but it seems to apply generally.

**VS Code extension or command-line**
command-line pyright 1.1.382

**Comment by erictraut** [comment]
A `py.typed` is applicable only for packages installed within your Python environment. It has no meaning (and is ignored by pyright) for files within your project. 

In your example, the files you are type checking are all under your project root, so they're all treated as files within your project, but you're also including the `example-py-pkg` in your `PYTHONPATH` environment variable, so it is also treated as an installed package. This is confusing pyright, and the behavior you're seeing here is a result.

If you want pyright to treat `example-py-pkg` as an installed package, you can do an editable install and then exclude it from type checking specifying an "exclude" in your configuration. If you want pyright to treat `example-py-pkg` as part of your project, you can set the "extraPaths" to point to this directory so imports resolve correctly.

```
extraPaths = ["example-py-pkg"]
```

Either of these approaches will fix the issue you're seeing.
